src/net_processing.cpp

```diff
@@ -2817,23 +2833,39 @@ void ProcessMessage(
         const CTransaction& tx = *ptx;
 
         const uint256& txid = ptx->GetHash();
-        pfrom.AddInventoryKnown(txid);
+        const uint256& wtxid = ptx->GetWitnessHash();
 
         LOCK2(cs_main, g_cs_orphans);
 
+        CNodeState* nodestate = State(pfrom.GetId());
+
+        const uint256& hash = nodestate->m_wtxid_relay ? wtxid : txid;
+        pfrom.AddInventoryKnown(hash);
+
         TxValidationState state;
 
-        CNodeState* nodestate = State(pfrom.GetId());
-        nodestate->m_tx_download.m_tx_announced.erase(txid);
-        nodestate->m_tx_download.m_tx_in_flight.erase(txid);
-        EraseTxRequest(txid);
+        nodestate->m_tx_download.m_tx_announced.erase(hash);
+        nodestate->m_tx_download.m_tx_in_flight.erase(hash);
+        EraseTxRequest(hash);
 
         std::list<CTransactionRef> lRemovedTxn;
 
-        if (!AlreadyHave(CInv(MSG_TX, txid), mempool) &&
+        // We do the AlreadyHave() check using wtxid, rather than txid - in the
+        // absence of witness malleation, this is strictly better, because the
+        // recent rejects filter may contain the wtxid but will never contain
+        // the txid of a segwit transaction that has been rejected.
+        // In the presence of witness malleation, it's possible that by only
+        // doing the check with wtxid, we could overlook a transaction which
+        // was confirmed with a different witness, or exists in our mempool
+        // with a different witness, but this has limited downside:
+        // mempool validation does its own lookup of whether we have the txid
+        // already; and an adversary can already relay us old transactions
+        // (older than our recency filter) if trying to DoS us, without any need
+        // for witness malleation.
+        if (!AlreadyHave(CInv(MSG_WTX, wtxid), mempool) &&
```

Comment:
"spuriously re-download transactions and only gets one version in your mempool" already happens with RBF; I don't think this is meaningfully different?

(If you're just varying the witness you don't need SIGHASH_NOINPUT, though, so I think there's a good chance the txid would be different too for eltoo transactions if that comes into play)

Reply:
I don't understand the comment here -- are you saying wtxid-relay is somehow incompatible with a future change to the protocol?  It seems to me that (a) wtxid-relay is more likely to support the ability to relay transactions with same underlying txid but different wtxid than our existing relay protocol, and (b) whether we adopt changes to the mempool acceptance rules to allow for this is otherwise orthogonal to this PR.  

If you're asking for a change here, can you indicate what that is?

Reply:
> "spuriously re-download transactions and only gets one version in your mempool" already happens with RBF; I don't think this is meaningfully different?

I think you're right we do already have this kind of case due to single-party RBF. After thoughts, qualifying them of _spurious_ was a fallacy as a smaller valid witness may increase transaction feerate and such increase its probabilities of propagating and confirming. So we _have to_ download them anyway.

(I agree SIGHASH_NOINPUT doesn't force you to vary witness, it's more likely we would do it, fyi another [generalized channel](https://eprint.iacr.org/2020/476.pdf) proposal leveraging per-party adaptor sigs to get symmetric states)

I'm not advocating on any change here, I was more brooding on the fact that fancier on-top-of-Bitcoin protocols and automatic dynamic-fee bumping as currently gauged by LN implementations will have for side-effect to increase bandwidth demand in case of mempool congestion.

> and (b) whether we adopt changes to the mempool acceptance rules to allow for this is otherwise orthogonal to this PR.

Yes, I was wrong that's orthogonal to tx-relay, it's more being sure that these potential bandwidth demand surges are priced "correctly" with regards to BIP 125 rule 4.